resolver/minaddr.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resolver
+
+import (
+	"context"
+	"io"
+)
+
+// MinAddresses decorates the given resolver so it sends a set of addresses that has at
+// least as many entries as the given minimum. If the given resolver provides a smaller
+// set of addresses, it replicates those addresses until the give minimum is reached.
+//
+// This will cause a client to effectively make redundant connections to the same address
+// which is particularly useful when the addresses are virtual IPs, which  actually have
+// multiple servers behind them. This is appropriate for environments like Kubernetes
+// (which uses virtual IPs for non-headless services) and for use with services that have
+// hardware/cloud load balancers in front.
+//
+// To avoid "hot spotting", where one backend address gets more load than others, this
+// always fully replicates the set. So it will always report at least minAddresses, but
+// could report nearly twice as many: in the case where the set from the underlying
+// resolver has minAddresses-1 entries, this will provide (minAddresses-1)*2 entries.
+func MinAddresses(other Resolver, minAddresses int) Resolver {
```

Comment:
I think the name MinAddresses is a little bit opaque, and I wonder if there's potentially a better way to phrase this. `RepeatAddresses`? `MinConnections`? Nothing really quite makes it obvious, which is frustrating.

Reply:
I like `MinConnections`. That is at least a little more clear about the _impact_ of this, since it obviously cannot fabricate distinct IP addresses but rather causes the client to establish multiple connections to the same IP address.